### DISCUSSION TITLE
Enable Addons to create 3D claims

### DISCUSSION THREAD

**Discussion by Jikoo** [GriefPrevention/GriefPrevention#1146] [Ideas]
My vision:
* GP provides its own claim creation system that addons can override/modify
  * GP claims are 2 points, protect area up to build height
  * Claims may not overlap
  * Can use `Integer.MAX_VALUE` for max Y for futureproofing instead of build height
* Addons can provide whatever claim creation system they like
  * Addons can create fully 3D claims
  * Addons can create overlapping claims (AKA subclaims)
  * Addons can set the world on fire and watch it burn
* GP will use claim bounding box for all containment/conflict checks
  * Mostly already does since #1126
  * Ideal future state: Claim extends BoundingBox

There is very little GP code that actually requires GP to have a concept of 2D vs 3D claims at all. The only thing I can think of off the top of my head that isn't claim creation is piston movement, and even that is such a minimal optimization that we could likely remove it without seeing much real impact - it only affects pistons moving blocks directly in line up and down.

Clearly per https://github.com/TechFortress/GriefPrevention/pull/1145#issuecomment-742689472 this doesn't line up with your vision:
> GP will never have a concept of a max Y in itself. Plan to let addons implement that. Probably a v19 thing to refactor claims (which would include removing subclaims)

Could you clarify how you want to make #269 possible? I get that with the addition of the ClaimPermissionCheckEvent it would be possible to override GP's behavior completely and ignore existing claims, but without a concept of max Y an addon would basically have to override both claiming and the datastore.

**Comment by Jikoo** [comment]
I have a pretty decent idea of how I'd implement things if you want clarification on any points, I just don't want to spend like 30 minutes writing pseudocode for project structure if you're not interested.

**Reply by RoboMWM** [reply]
Currently not in the purview of focus until v19 or so, therefore here lies my offhand and unrefined remarks: GP will not be storing/worrying about max height. I get it though, there is "greaterBoundaryCorner" so maybe it's already there, so maybe I could look at that (later), but logic-wise GP itself will ignore max height in its own code.

**Reply by Jikoo** [reply]
That's fine, just wanted clarification so I don't waste effort on something that isn't desirable - it would suck to actually write claims to extend bounding boxes only to have it declined because we're thinking in different directions. I'll ping you on this if/when I get to that point.
